beater/authorization/anonymous.go

```diff
@@ -19,12 +19,41 @@ package authorization
 
 import (
 	"context"
+	"fmt"
 )
 
-// AnonymousAuth implements the Authorization interface.
-type AnonymousAuth struct{}
+func newAnonymousAuth(allowAgent, allowService []string) *anonymousAuth {
+	a := &anonymousAuth{
+		allowedAgents:   make(map[string]bool),
+		allowedServices: make(map[string]bool),
+	}
+	for _, name := range allowAgent {
+		a.allowedAgents[name] = true
+	}
+	for _, name := range allowService {
+		a.allowedServices[name] = true
+	}
+	return a
+}
+
+// anonymousAuth implements the Authorization interface, allowing anonymous access with
+// optional restriction on agent and service name.
+type anonymousAuth struct {
+	allowedAgents   map[string]bool
+	allowedServices map[string]bool
+}
 
-// AuthorizedFor always returns a Result indicating the request is authorized and anonymous.
-func (AnonymousAuth) AuthorizedFor(context.Context, Resource) (Result, error) {
-	return Result{Anonymous: true, Authorized: true}, nil
+// AuthorizedFor always returns a Result indicating the request is anonymous, and
+// authorized if the resource is permitted for anonymous access.
+func (a *anonymousAuth) AuthorizedFor(ctx context.Context, resource Resource) (Result, error) {
+	result := Result{Authorized: true, Anonymous: true}
+	switch {
+	case resource.AgentName != "" && len(a.allowedAgents) > 0 && !a.allowedAgents[resource.AgentName]:
+		result.Authorized = false
+		result.Reason = fmt.Sprintf("agent %q not allowed", resource.AgentName)
+	case resource.ServiceName != "" && len(a.allowedServices) > 0 && !a.allowedServices[resource.ServiceName]:
+		result.Authorized = false
+		result.Reason = fmt.Sprintf("service %q not allowed", resource.ServiceName)
```

Comment:
Indeed, it's by design:
 - the initial auth check for requests uses `Resource{}`, to check if the client has _some_ access before agent/service is known
 - for some endpoints (one endpoint -- agent config) the agent name isn't known, only service name is known

To prevent agents from bypassing the allowed agent/service checks we rely on agent and service name always being set where it matters. In agent config the service name is always set, and in intake both the agent name and service name are always set.

TBH I don't really feel 100% comfortable with this, so I'm open to alternatives.

Reply:
That's indeed a tricky one!
How about we set a `any` (or `*` or something else) as agent name for the initial check, and give it a special meaning in the auth check, but ensure that this value is removed in case it is set by the requester?

> for some endpoints (one endpoint -- agent config) the agent name isn't known, only service name is known

Reply:
Another more radical option: introduce two different methods, Authenticate and Authorize.

Authenticate checks that a provided secret token or API Key is valid, or (if no auth token is provided) that anonymous access is allowed. If API Key auth is used, the API Key ID will be obtained so we can associate it with events.

If an agent authenticates, we assume it is allowed to send events and query agent config and we'll proceed to decoding events and parsing agent config queries. This is effectively like AuthorizedFor(Resource{}) succeeding. Once authenticated, Authorize will be used to check that that the agent is allowed to perform an action for a given agent name and/or service name.
 
The challenge with this is that we currently have the ability to create API Keys for sending events, querying agent config, and indexing source maps. Since source map uploading is moving to Kibana, I think we could simplify things by saying that if an agent has an API Key with _some_ apm application privileges, then it's good enough for agent authentication.

Reply:
I like that. It would generally be cleaner to distinguish between authentication and authorization.